### PR TITLE
fix(task): preserve inner quotes for cmd /c tasks and hooks on windows

### PR DESCRIPTION
**Pull request by JamBalaya56562**
## Problem

On Windows, inner double quotes in a task's `run` string — and in `[hooks]` commands — are mangled when spawned through the default `cmd /c` shell. Reported in #9355 (originally #6119).

```toml
[tasks.repro]
run = 'uv run --no-project python -c "import pathlib; print(pathlib.Path.home())"'
```

```
DEBUG $ cmd /c uv run --no-project python -c "import pathlib; print(pathlib.Path.home())"
  File "<string>", line 1
    "import
SyntaxError: unterminated string literal
```

The debug line looks correct, but Python receives just `"import`. Likewise `run = 'echo "x"'` prints `\"x\"`. `[hooks]` (`postinstall`, `enter`, `cd`, …) are hit identically, and unlike tasks they have **no** `usage`/shebang/file-based escape hatch.

## Root cause

mise builds the argv `["cmd", "/c", <script>, …]` and passes the script as an ordinary `std::process::Command` argument. On Windows, std serializes argv into `lpCommandLine` using MSVCRT/`CommandLineToArgvW`-style quoting (it wraps the script in quotes and escapes inner `"` as `\"`). `cmd.exe` does **not** parse that convention — it treats `\"` literally — so the child receives mangled arguments. No `raw_arg` was used anywhere in the tree.

## Fix

When the inline/hook shell is `cmd.exe`, build the command line **verbatim** via raw command-line args:

- The whole command is wrapped in a single outer quote pair and run as `cmd /s /c "<body>"`. `/s` strips exactly that one outer pair and leaves the remainder — including the user's inner quotes — untouched.
- Forwarded args are MSVCRT-quoted **inside** the outer pair so the program still sees them as separate arguments, preserving the spaces-in-args fix from #6744 (e.g. `mise run type ".\test dir\file.txt"`).
- `[hooks]` spawn through duct, which can't emit raw args, so the cmd case now spawns a `std::process::Command` directly and forwards stdout to stderr to match the previous `stdout_to_stderr()` behavior.
- Non-cmd shells (`pwsh -Command`, `bash -c`, …) are gated out via `is_cmd_shell_program` and keep the existing behavior.

New helpers live in `src/path.rs` (`is_cmd_shell_program`, `cmd_verbatim_args`, `quote_arg_for_cmd_body`) so the other call sites can reuse them later.

## Scope / follow-ups

- This PR fixes **inline tasks** and **`[hooks]`** (the two paths reported in #9355).
- The same root cause also affects `watch_files`, `deps`, `{{ exec() }}`, `mise exec -c`, and backend scripts. Those can reuse `cmd_verbatim_args` in a follow-up.
- The separate `shell = "bash -c"` `$0`-shift issue noted in the discussion is left for a follow-up.

## Testing

- Unit tests for `is_cmd_shell_program`, `cmd_verbatim_args`, and `quote_arg_for_cmd_body` (covering the MSVCRT quoting edge cases and the `type ".\test dir\file.txt"` case from `e2e-win/task_args.Tests.ps1`). I verified this pure string logic locally.
- New `e2e-win/task_quoting.Tests.ps1`: asserts inner quotes survive (no `\` mangling) and that an inner-quoted `-c` argument reaches the program as a single argument.
- Heads-up: my local environment couldn't compile the full crate (a C-dependency toolchain limitation) and can't compile the `#[cfg(windows)]` branches at all, so I'm **relying on CI** for the Windows build, clippy, and the `e2e-win` run.

Refs #9355, #6119.

---
*This pull request was prepared with the help of an AI coding assistant.*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed Windows task and hook execution so inline commands run via cmd preserve inner double quotes and don’t produce stray backslashes, improving reliability of quoted arguments.

* **Tests**
  * Added a Windows end-to-end test that validates quoted-string preservation across task runs (skips if Node is unavailable).

### CODE REVIEW

File: src/path.rs [hunk 1]
```diff
@@ -152,6 +152,85 @@ pub fn is_posix_shell_program(program: &Path) -> bool {
...
+    out
+}
+
+/// MSVCRT/`CommandLineToArgvW`-style quoting for a single argument, matching the
+/// rules `std::process::Command` uses on Windows. Used for forwarded args placed
+/// inside [`cmd_verbatim_args`]' outer quote pair so the *child program* (parsed
+/// by the C runtime) sees each as one argument. Only quotes when needed (empty,
+/// or contains space/tab/`"`); backslashes are doubled only where they precede a
+/// `"`.
+#[cfg_attr(not(windows), allow(dead_code))]
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply src/path.rs:208]: Good call — made `quote_arg_for_cmd_body` `pub(crate)` (02bf972) so the planned follow-ups (`watch_files`, `deps`, `{{ exec() }}`, `mise exec -c`) can reuse it without re-opening `path.rs`. Kept it `pub(crate)` rather than `pub` since mise is a binary crate with no external API surface.

*This reply was generated by an AI coding assistant.*

File: src/hooks.rs [hunk 2]
```diff
@@ -583,6 +583,50 @@ async fn execute(
...
+            c.envs(env.iter());
+            c.stdout(std::process::Stdio::piped());
+            let mut child = c.spawn()?;
+            let copier = child.stdout.take().map(|mut out| {
+                std::thread::spawn(move || {
+                    let mut err = std::io::stderr();
+                    let _ = std::io::copy(&mut out, &mut err);
+                })
+            });
+            let status = child.wait()?;
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply src/hooks.rs]: Thanks — I looked into this but don't believe it's a regression, so I left it as-is.

`with_pass_signals()` is on the *task* execution path (`CmdLineRunner` in `exec_program`). The hook path has always spawned through duct (`crate::cmd::cmd(...).run()`), which doesn't forward signals either, so the new cmd branch matches the prior hook behavior. Also, on Windows a console Ctrl+C is delivered by the OS to every process attached to the console (including the spawned `cmd` child), so the child receives it without explicit forwarding.

Happy to revisit if explicit handling is preferred.

*This reply was generated by an AI coding assistant.*

File: e2e-win/task_quoting.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+run = 'echo "My Great Document.typ"'
+
+[tasks.node_inner_quotes]
+run = 'node -e "console.log(2 + 2)"'
+'@ | Out-File -FilePath "mise.toml" -Encoding utf8NoBOM
+    }
+
+    AfterAll {
+        Set-Location $originalPath
+        Remove-Item -Path Env:\MISE_TRUSTED_CONFIG_PATHS -ErrorAction SilentlyContinue
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply e2e-win/task_quoting.Tests.ps1]: Done (02bf972). `BeforeAll` now captures `$originalTrustedConfigPaths` and `AfterAll` restores it (or removes it only if it wasn't set originally), matching the save/restore pattern the Cygwin test in `task.Tests.ps1` already uses.

*This reply was generated by an AI coding assistant.*

File: src/hooks.rs [hunk 4]
```diff
@@ -583,6 +583,50 @@ async fn execute(
...
+            if let Some(h) = copier {
+                let _ = h.join();
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply src/hooks.rs]: Fixed (02bf972). Replaced the piped-stdout + copier-thread shim with a direct descriptor redirect: the child's stdout is now a clone of mise's stderr handle (`std::io::stderr().as_handle().try_clone_to_owned()?`). There's no reader thread to join and no pipe EOF to wait on, so a hook that spawns a background child holding the write end can't block us, and stdout/stderr ordering is preserved. This matches the duct `stdout_to_stderr()` semantics the non-cmd path uses.

*This reply was generated by an AI coding assistant.*

File: src/path.rs [hunk 5]
```diff
@@ -152,6 +152,85 @@ pub fn is_posix_shell_program(program: &Path) -> bool {
...
+fn quote_arg_for_cmd_body(arg: &str) -> String {
+    if !arg.is_empty() && !arg.contains([' ', '\t', '"']) {
+        return arg.to_string();
+    }
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply src/path.rs]: Fixed (02bf972). `quote_arg_for_cmd_body` now also quotes args containing cmd metacharacters (`& | < > ( ) ^`), so a forwarded `a&b` reaches the child as a single argv value instead of being parsed by cmd after the outer `/s` quote pair is stripped. Added unit cases for `a&b`, `foo|bar`, `a>b`.

I intentionally left `%` out: cmd expands `%VAR%` even inside double quotes, so quoting can't protect it (that would need a separate approach).

*This reply was generated by an AI coding assistant.*
